File: api/app/schemas/catalog.py
"""Catalog: stores, products, variants, images, reviews."""

from __future__ import annotations

from datetime import datetime

from pydantic import Field

from app.schemas.enums import ProductStatus, StoreStatus
from app.schemas.envelope import CamelModel


class StoreSummary(CamelModel):
    """Store as shown on product cards / detail headers."""

    id: str
    name: str
    slug: str
    location: str | None = None
    status: StoreStatus


class StoreDetail(StoreSummary):
    """Full store detail (storefront page)."""

    bio: str | None = None
    created_at: datetime


class VariantOut(CamelModel):
    """A purchasable variant with price + live stock signal."""

    id: str
    sku: str
    options: dict[str, object] = Field(default_factory=dict)
    price_minor: int = Field(ge=0)
    currency: str = "USD"
    is_active: bool = True
    # Derived from inventory (qty_on_hand - qty_reserved). Boolean keeps stock
    # levels private while still driving the out-of-stock UI state.
    in_stock: bool = True
    restock_eta_days: int | None = None


class ProductImageOut(CamelModel):
    """A product image."""

    id: str
    url: str
    alt: str = ""
    position: int = 0
    variant_id: str | None = None


class ReviewOut(CamelModel):
    """A single product review."""

    id: str
    product_id: str
    user_id: str
    rating: int = Field(ge=1, le=5)
    title: str | None = None
    body: str = ""
    created_at: datetime


class ReviewCreate(CamelModel):
    """POST /products/{id}/reviews — one per (product, user)."""

    rating: int = Field(ge=1, le=5)
    title: str | None = Field(default=None, max_length=200)
    body: str = Field(default="", max_length=4000)


class ProductSummary(CamelModel):
    """Product as shown in lists / search results / recommendation cards."""

    id: str
    title: str
    slug: str
    category: str | None = None
    status: ProductStatus
    store: StoreSummary
    # Cheapest active variant price, for list/card display.
    from_price_minor: int | None = Field(default=None, ge=0)
    currency: str = "USD"
    primary_image: ProductImageOut | None = None
    # DERIVED rollups (recomputed on review write); rating_avg null until first review.
    rating_avg: float | None = Field(default=None, ge=0, le=5)
    rating_count: int = 0


class ProductDetail(ProductSummary):
    """Full product detail page payload."""

    description: str = ""
    attributes: dict[str, object] = Field(default_factory=dict)
    variants: list[VariantOut]
    images: list[ProductImageOut]
    created_at: datetime
    updated_at: datetime

File: web/src/lib/adapters/catalog.ts
/**
 * Contract → view-model adapters (Day-22 flip-to-live, ADR-0040).
 *
 * The live backend returns NESTED catalog shapes (`product.store.name`,
 * `product.from_price_minor`, `product.primary_image.alt`, …). The FE
 * view-models in `@/lib/api-types` are FLAT (`maker`, `price_cents`,
 * `image_alt`). These mappers live in ONE place so the route handlers and
 * server components stay thin and the components are unchanged.
 *
 * Money note: the FE field is still called `price_cents` (legacy name) but it
 * carries minor units — we set it straight from the backend `*_minor` value.
 */
import type {
  AgentRecommendation,
  ProductDetail,
  ProductImage,
  ProductVariant,
  SearchResult,
  StockState,
} from "@/lib/api-types";

/* ----------------------- backend (wire) shapes ----------------------- */

interface BackendStore {
  id: string;
  name: string;
  slug: string;
  location: string | null;
  status: string;
}

interface BackendImage {
  id: string;
  url: string;
  alt: string | null;
  position: number;
  variant_id: string | null;
}

interface BackendVariant {
  id: string;
  sku: string;
  options: Record<string, unknown>;
  price_minor: number;
  currency: string;
  is_active: boolean;
  in_stock: boolean;
  restock_eta_days: number | null;
}

/** `ProductSummary` on the wire — shared by search rows, recs, and detail. */
export interface BackendProductSummary {
  id: string;
  title: string;
  slug: string;
  category: string;
  status: string;
  store: BackendStore;
  from_price_minor: number;
  currency: string;
  stock: StockState;
  primary_image: BackendImage | null;
  rating_avg: number | null;
  rating_count: number;
}

/** `ProductDetail` on the wire — `ProductSummary` + the detail-only fields. */
export interface BackendProductDetail extends BackendProductSummary {
  description: string;
  attributes: Record<string, unknown>;
  variants: BackendVariant[];
  images: BackendImage[];
  created_at: string;
  updated_at: string;
}

/** A `GET /search` row: `{ product: ProductSummary, score }`. */
export interface BackendSearchRow {
  product: BackendProductSummary;
  score: number;
}

/** A live agent `done`-frame recommendation (nests `ProductSummary`). */
export interface BackendRecommendation {
  product: BackendProductSummary;
  reason: string;
}

/* ------------------------------ mappers ------------------------------ */

/** Flatten a wire `ProductSummary` → FE `SearchResult`. */
export function toSearchResult(p: BackendProductSummary): SearchResult {
  return {
    id: p.id,
    slug: p.slug,
    title: p.title,
    maker: p.store.name,
    price_cents: p.from_price_minor,
    currency: p.currency,
    stock: p.stock,
    image_alt: p.primary_image?.alt ?? null,
  };
}

/** `GET /search` → `SearchResult[]`. */
export function toSearchResults(rows: BackendSearchRow[]): SearchResult[] {
  return rows.map((row) => toSearchResult(row.product));
}

function toProductImage(img: BackendImage): ProductImage {
  return {
    id: img.id,
    alt: img.alt ?? "",
    // No `tone` from live — components fall back to a token placeholder when absent.
  };
}

function toProductVariant(v: BackendVariant): ProductVariant {
  // The wire carries an options dict; the FE renders a single name/value pair.
  const [optName, optValue] = Object.entries(v.options)[0] ?? ["", ""];
  return {
    id: v.id,
    option_name: String(optName),
    option_value: String(optValue ?? ""),
    stock: v.in_stock ? "in_stock" : "out_of_stock",
  };
}

/** `GET /products/{idOrSlug}` → FE `ProductDetail`. */
export function toProductDetail(p: BackendProductDetail): ProductDetail {
  return {
    id: p.id,
    slug: p.slug,
    title: p.title,
    maker: p.store.name,
    maker_location: p.store.location,
    price_cents: p.from_price_minor,
    currency: p.currency,
    stock: p.stock,
    description: p.description,
    rating: p.rating_avg,
    review_count: p.rating_count,
    images: p.images.map(toProductImage),
    variants: p.variants.map(toProductVariant),
  };
}

/** A live agent recommendation → FE `AgentRecommendation` (same flatten as search). */
export function toAgentRecommendation(
  r: BackendRecommendation,
): AgentRecommendation {
  const p = r.product;
  return {
    product_id: p.id,
    slug: p.slug,
    title: p.title,
    maker: p.store.name,
    price_cents: p.from_price_minor,
    currency: p.currency,
    stock: p.stock,
    reason: r.reason,
    image_alt: p.primary_image?.alt ?? null,
  };
}
